### PR TITLE
add a Json Pointer parser

### PR DESCRIPTION
**Pull request by jbtrystram**
It's needed to properly finish https://github.com/eclipse/hono/pull/2009/ :)

Signed-off-by: Jean-Baptiste Trystram <jbtrystram@redhat.com>

### CODE REVIEW

File: core/src/main/java/org/eclipse/hono/util/JsonHelper.java [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2019 Contributors to the Eclipse Foundation
+ * Copyright (c) 2019-2020 Contributors to the Eclipse Foundation
```
Comment (sophokles73) [inline core/src/main/java/org/eclipse/hono/util/JsonHelper.java:2]: 2019, 2020

File: core/src/main/java/org/eclipse/hono/util/JsonHelper.java [hunk 2]
```diff
@@ -55,4 +58,100 @@ private JsonHelper() {
...
+                    clazz.getSimpleName());
+            return defaultValue;
+        }
+    }
+
+    /**
+     * Return the value designated from a JsonPath in the given JsonObject.
+     *
+     * @param jsonObject The JSON object to get the value from.
+     * @param path The JsonPath representation of the value to extract.
```
Comment (sophokles73) [inline core/src/main/java/org/eclipse/hono/util/JsonHelper.java:101]: FMPOV we need a definition of the path syntax. Is there an existing definition somewhere online which we can refer to?

**Comment by dejanb** [comment]
Looks good to me in general. I agree with Kai, that it'd be good to have a link to the "spec" (https://goessner.net/articles/JsonPath/)

Another question is, why we don't use the existing library for it? https://github.com/json-path/JsonPath

**Comment by sophokles73** [comment]
> Another question is, why we don't use the existing library for it?

@dejanb I had the same feeling but that doesn't help much because we are not trying to extract data from an existing JSON document but instead simply want to use the notation in order to specify a query on the underlying persistence technology.
BTW MongoDB doesn't support queries expressed in JSONPath even though its data format is already JSON :-(
Another alternative to using a subset of JSONPath might be to use [JSON Pointer](https://tools.ietf.org/html/rfc6901) instead which is much more lightweight and only specifies how to address a particular value within a document ...

**Comment by jbtrystram** [comment]
@sophokles73 I was actually just looking at the Json Pointer spec and thinking it may be a better fit for that use case.

**Comment by dejanb** [comment]
@sophokles73 @jbtrystram I see. I'm OK with exploring JSON Pointer solution as well.

**Comment by jbtrystram** [comment]
@dejanb @sophokles73 I updated this PR with a Json Pointer implementation. I added a bunch of tests for the special use cases and as far as I know it's compliant to the spec. 
Let me know what you think.

**Comment by sophokles73** [comment]
@jbtrystram what about `io.vertx.core.json.pointer.JsonPointer`? Doesn't that work?

**Comment by jbtrystram** [comment]
:( damn, this was added in vertx after this [issue](https://github.com/eclipse-vertx/vert.x/issues/1661) stated they won't support it.
This can be closed I guess

**Pull request closed without merging**